proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,314 @@
+
+      -- desugars to `("Example " ++) . interpolatePrec 0 foo $ ""`
+      original :: String
+      original = [s|Example ${foo}|]
+
+      -- desugars to `("Example " ++) . MyInterpolate.interpolatePrec 0 foo $ ""`
+      custom :: String
+      custom = [s.MyInterpolate|Example ${foo}|]
+
+* Generalize to any output type
+
+  * Break out a type class representing "Type that an interpolation can result in" and enable specifying any module defining a compatible interface, similar to ``QualifiedDo`` (would also enable swapping out the ``Interpolate`` type class, like the previous bulletpoint). e.g.
+    ::
+
+      {----- built-in to GHC -----}
+
+      class Interpolate s where
+        interpolateRaw :: String -> s -> s
+      class Interpolate s => InterpolateValue s a where
+        interpolatePrec :: Int -> a -> s -> s
+
+      instance Interpolate String where
+        interpolateRaw = (++)
+
+      -- the usual instances
+      instance InterpolateValue String String
+      instance InterpolateValue String Int
+
+      {----- user-defined -----}
+
+      module MySqlQuery where
+
+      data SqlQuery = SqlQuery { queryText :: String, queryVals :: [SqlVal] }
+      instance Interpolate SqlQuery where
+        interpolateRaw q1 (SqlQuery q2 vs) = SqlQuery (q1 <> q2) vs
+      instance InterpolateValue SqlQuery String where
+        interpolatePrec _ s (SqlQuery text vs) = SqlQuery text (SqlString s : vs)
+      instance InterpolateValue SqlQuery Int where
+        interpolatePrec _ x (SqlQuery text vs) = SqlQuery text (SqlInt x : vs)
+
+      module Main where
+
+      -- desugars to `interpolateRaw "SELECT * FROM foo WHERE id = " . interpolatePrec 0 id $ ""`
+      pwnable :: Int -> String
+      pwnable id = [s|SELECT * FROM foo WHERE id = ${id}|]
+
+      -- desugars to `MySqlQuery.interpolateRaw "SELECT * FROM foo WHERE id = " . MySqlQuery.interpolatePrec 0 id $ ""`
+      safe :: Int -> SqlQuery
+      safe id = [s.MySqlQuery|SELECT * FROM foo WHERE id = ${id}|]
+
+  * Benefit: one more metaprogramming technique, except this wouldn't require Template Haskell
+  * Drawback: requires ``MultiParamTypeClasses``, probably has worse type inference, not sure how to integrate ``OverloadedStrings``
+  * Perhaps this should just be left to third-party QuasiQuoters using something like ``ghc-meta``?
+
+* Different delimiter
+
+  * Could use ``f"`` like Python, with ``f`` for format. ``s`` for "String" seems a bit ad-hoc, but it does "look better" for some reason. ``s`` is also a bit better if the user forgets to enable ``-XStringInterpolation`` because ``f`` is a not-uncommon name for functions and ``f"asdf"``, being parsed as ``f "asdf"``, would work more often than ``s"asdf"`` would.
+  * Could reuse QuasiQuote syntax, e.g. ``[s|`` or ``[fmt|``, except it would be special and NOT use Template Haskell.
+
+* Different interpolation delimiter, e.g. ``#{foo}``
+
+  * Most languages use ``$``, and I see no reason to deviate
```

Comment:
Ruby uses `#{..}`, as do packages `interpolate` and `string-interpolate`. `neat-interpolation` uses `$`, and I think either alternative would be recognisable as interpolation.

Reply:
I did say "most" 😉. I agree, either would be recognizable. But `$` _is_ more common, no?

Reply:
I haven't sat down and counted, and I'm not certain what the correct class of things to count would be: common programming languages? Interpolation libraries on Hackage? Interpolation libraries on Hackage, weighted by revdeps in some way? I don't think it matters a whole lot.

Reply:
We used `#{..}` instead of `${..}` in the `str` quasiquoter in streamly (https://streamly.composewell.com/haddocks/streamly-core-0.1.0/Streamly-Unicode-String.html). The reason was that we were writing shell script snippets using this and `${..}` conflicted with shell's syntax. If we ignore that, the `$` syntax is perhaps more intuitive.

Reply:
Specifically for interpolating in shell contexts, sure. But I don't think `\${x}` is that terrible for this fraction of use cases. You already have to do similar in shell, e.g.

```bash
bash -c "PATH=/foo/bin:\${PATH} ${cmd}"
```

Reply:
maybe `\{..}` could also work? [idris2 uses this syntax](https://idris2.readthedocs.io/en/latest/reference/strings.html#interpolated-strings) and i think this is quite nice as it looks similar to lambda syntax `\x -> y`

Reply:
I'm not sure how "looking like a lambda" is relevant here. `s"User is \{name}"` isn't binding `name` like a lambda argument. Also, IMO it would be confusing for `\` to both denote/escape special characters and interpolate variables. I would also imagine this would be harder to implement in the parser.

I'd really like to limit bikeshedding this any further. Every suggestion is going to have tradeoffs. IMO the intuition/popularity of `${...}` greatly exceeds the cost of escaping it in specific scenarios[^1]. This assumption should be challenged first before we can discuss alternatives.

[^1]: In a sense, the two are correlated. The more popular a delimiter is, the more likely you'd need to escape it. So aiming to find a delimiter that's less likely to need escaping will necessarily be some bespoke delimiter no one else is using (yet).

Reply:
you're right. I'll be happy to (finally) have a string interpolation syntax regardless how does it look like.